Notebooks/Pluto/p10.jl

```diff
@@ -0,0 +1,129 @@
+### A Pluto.jl notebook ###
+# v0.12.17
+
+using Markdown
+using InteractiveUtils
+
+# ╔═╡ fe142e4e-46dd-11eb-3b8b-79a5e76789a5
+using Yao, YaoPlots
+
+# ╔═╡ fb8094ea-46da-11eb-3653-4b38fefb019c
+md"# qRAM and Uncomputation"
+
+# ╔═╡ 259b2164-46db-11eb-1969-19fbb8a1d18c
+md"Consider encoding data in qubits. Assume the following array."
+
+# ╔═╡ 99efc04a-46dd-11eb-1084-23f01b24d6e0
+a = [0, 0, 1, 0, 0]
+
+# ╔═╡ cdab262c-46dd-11eb-1d74-d579e68851ee
+md"It'll take 5 qubits to encode this."
+
+# ╔═╡ dc1b6eec-46dd-11eb-324e-5dc8a639b138
+ArrayReg(bit"00100") |> r->measure(r)
+
+# ╔═╡ 0e7ff754-46de-11eb-13d5-35b80cd84409
+md"or"
+
+# ╔═╡ 160c0030-46de-11eb-0970-593c919558e4
+zero_state(5) |> put(5, 3=>X) |> r->measure(r)
+
+# ╔═╡ 4eeb3288-46de-11eb-18d5-59ff9c23a0b0
+md"Either way, lets assume we've to encode 4 such arrays, in qubits."
+
+# ╔═╡ 67c39c3a-46de-11eb-009f-9396ff599539
+b = [0, 1, 1, 0, 1]
+
+# ╔═╡ 868c2a06-46de-11eb-1b0b-f15ec79de048
+c = [1, 1, 0, 0, 0]
+
+# ╔═╡ be7b3a6a-46de-11eb-1ecd-4b300e1c4735
+d = [1, 0, 1, 1, 1]
+
+# ╔═╡ 95399610-46de-11eb-3876-253ee5fea025
+md"To encode these 4 arrays, it'd take 20 qubits, judging by the above approach. But using QRAMS, we can use 7 qubits, to encode all the 4 arrays."
+
+# ╔═╡ f7efa09c-46de-11eb-293a-e18468e3854d
+let
+	f(x) = chain(7, [control(1:2, (k+2)=>X) for k in findall(isone, x)])
+	global QRAM = chain(7, repeat(H, 1:2), repeat(X, 1:2), f(a), repeat(X, 1:2), put(1=>X), f(b), put(1=>X), put(2=>X), f(c), put(2=>X), f(d))
```

Comment:
I'm not sure `global` is necessary? maybe just use `begin ... end`? I think it would be nice if we don't ask our readers to learn something extra and not common.

Reply:
You're right, I'll change that
I apologize, thats an old habit😂